src/librustc_data_structures/bitvec.rs

```diff
@@ -15,26 +15,193 @@ pub struct BitVector {
 
 impl BitVector {
     pub fn new(num_bits: usize) -> BitVector {
-        let num_words = (num_bits + 63) / 64;
+        let num_words = u64s(num_bits);
         BitVector { data: vec![0; num_words] }
     }
 
-    fn word_mask(&self, bit: usize) -> (usize, u64) {
-        let word = bit / 64;
-        let mask = 1 << (bit % 64);
-        (word, mask)
-    }
-
     pub fn contains(&self, bit: usize) -> bool {
-        let (word, mask) = self.word_mask(bit);
+        let (word, mask) = word_mask(bit);
         (self.data[word] & mask) != 0
     }
 
     pub fn insert(&mut self, bit: usize) -> bool {
-        let (word, mask) = self.word_mask(bit);
+        let (word, mask) = word_mask(bit);
         let data = &mut self.data[word];
         let value = *data;
         *data = value | mask;
         (value | mask) != value
     }
+
+    pub fn insert_all(&mut self, all: &BitVector) -> bool {
+        assert!(self.data.len() == all.data.len());
+        let mut changed = false;
+        for (i, j) in self.data.iter_mut().zip(&all.data) {
+            let value = *i;
+            *i = value | *j;
+            if value != *i { changed = true; }
+        }
+        changed
+    }
+
+    pub fn grow(&mut self, num_bits: usize) {
+        let num_words = u64s(num_bits);
+        let extra_words = self.data.len() - num_words;
+        self.data.extend((0..extra_words).map(|_| 0));
+    }
+}
+
+/// A "bit matrix" is basically a square matrix of booleans
+/// represented as one gigantic bitvector. In other words, it is as if
+/// you have N bitvectors, each of length N.
+#[derive(Clone)]
+pub struct BitMatrix {
+    elements: usize,
+    vector: Vec<u64>,
```

Comment:
@pnkfelix 

> I don't know the BitVector API, so maybe there's an obvious reason not to do this, but: Did you consider implementing this as a BitVector ?

I'm not sure if you mean "on top of a BitVector" (i.e., instead of a `Vec<u64>`), or "by adding APIs to `BitVector`". I considered both: I rejected the latter because it seemed confusing to have some methods that treat the list of bits as a single list, and another that treat them as a square matrix. I decided against building on bitvector because there is so little code and because I'd have to at minimum add some methods to bitvector for grabbing a slice of `u64` -- but I guess they could be private to this module. Seems like it might be ok, not sure how much of an improvement it would be.

(And yes I wouldn't want to fragment entries across u64s, not worth it.)